mlc_llm/relax_model/llama_batched_vllm.py

```diff
@@ -124,18 +126,43 @@ def forward(
             )
         else:
             # Decode, using vLLM kernel
+            exp_sums = nn.emit(
+                relax.op.builtin.alloc_tensor(
+                    relax.ShapeExpr((num_tokens, self.num_query_heads, self.max_num_partitions)),
+                    dtype="float32",
+                    runtime_device_index=0,
+                )
+            )
+            max_logits = nn.emit(
+                relax.op.builtin.alloc_tensor(
+                    relax.ShapeExpr((num_tokens, self.num_query_heads, self.max_num_partitions)),
+                    dtype="float32",
+                    runtime_device_index=0,
+                )
+            )
+            tmp_out = nn.emit(
+                relax.op.builtin.alloc_tensor(
+                    relax.ShapeExpr(
+                        (num_tokens, self.num_query_heads, self.max_num_partitions, self.head_dim)
+                    ),
+                    dtype=queries.struct_info.dtype,
+                    runtime_device_index=0,
+                )
+            )
```

Comment:
We don't need these allocs for the V1 path right? Why not allocate them in `single_query_cached_kv_attention_v2` in your cpp code? That way, we don't need to worry about the alloc overhead for the v1 path, which should be more frequent. 

And for the V2 path, given that we don't have a good allocator yet, I wonder which of
* Using the v1 kernel even if the v2 condition is met `max_context_len <= 8192 && (max_num_partitions == 1 || num_seqs * num_heads > 512);`
* Using the v2 with the alloc overhead
is expected to be faster. Do you have any thought on that? I think it is entirely ok to disable v2 until we land an improved allocator.

Reply:
From the numbers I share last week, when the v2 condition is met, it is faster even though there's overhead of allocation. I didn't do allocation in cpp because I'd like the memory being managed by memory planner. I think one optimization we can do before the better allocator is doing a global allocation that is reused across attention layers, we can make it a static member of `LlamaPagedAttention`